File: java/ToolGood/Words/StringSearch.java
package ToolGood.Words;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import java.util.Map;

import org.springframework.util.StringUtils;

import ToolGood.Words.internals.TrieNode;

public class StringSearch {
    protected TrieNode[] _first = new TrieNode[Character.MAX_VALUE + 1];


    public void SetKeywords(List<String> _keywords)
    {
        TrieNode[] first =  new TrieNode[Character.MAX_VALUE + 1];
        TrieNode root = new TrieNode();

        _keywords.forEach(p->{
            if(StringUtils.isEmpty(p)==false){ 
                TrieNode nd = _first[ p.charAt(0)];
                if (nd == null) {
                    nd = root.Add(p.charAt(0));
                    first[p.charAt(0)] = nd;
                }
                for (int i = 1; i < p.length(); i++) {
                    nd = nd.Add(p.charAt(i));
                }
                nd.SetResults(p);
            }
        });
 
        this._first = first;// root.ToArray();

        Map<TrieNode, TrieNode> links = new Hashtable<TrieNode, TrieNode>();
        root.m_values.forEach((k,v)->{
            TryLinks(v, null, links);
        });
        links.forEach((k,v)->{
            k.Merge(v, links);
        });
    }

    private void TryLinks(TrieNode node, TrieNode node2, Map<TrieNode, TrieNode> links)
    {
        node.m_values.forEach((Key,Value)->{
            TrieNode tn = null;
            if (node2 == null) {
                tn = _first[Key];
                if (tn != null) {
                    links.put(Value, tn);
                }
            } else if (node2.TryGetValue(Key, tn)) {
                links.put(Value, tn);
            }
            TryLinks(Value, tn, links);
        });
    }


    public String FindFirst(String text)
    {
        TrieNode ptr = null;
        for (int i = 0; i < text.length(); i++) {
            Character t=text.charAt(i);
            TrieNode tn=null;
            if (ptr == null) {
                tn = _first[t];
            } else {
                if (ptr.TryGetValue(t, tn) == false) {
                    tn = _first[t];
                }else {
                    tn=ptr.GetValue(t);//TODO: 不太清楚 java out参数 怎么做？ 
                }
            }
            if (tn != null) {
                if (tn.End) {
                    return tn.Results.get(0);
                }
            }
            ptr = tn;
        }
        return null;
    }

    public List<String> FindAll(String text)
    {
        TrieNode ptr = null;
        List<String> list = new ArrayList<String>();

        for (int i = 0; i < text.length(); i++) {
            Character t=text.charAt(i);
            TrieNode tn=null;
            if (ptr == null) {
                tn = _first[t];
            } else {
                if (ptr.TryGetValue(t, tn) == false) {
                    tn = _first[t];
                }else {
                    tn=ptr.GetValue(t);//TODO: 不太清楚 java out参数 怎么做？ 
                }
            }
            if (tn != null) {
                if (tn.End) {
                    tn.Results.forEach(item->{
                        list.add(item);
                    });
                }
            }
            ptr = tn;
        }
        return list;
    }

    public boolean ContainsAny(String text)
    {
        TrieNode ptr = null;
        for (int i = 0; i < text.length(); i++) {
            Character t=text.charAt(i);
            TrieNode tn=null;
            if (ptr == null) {
                tn = _first[t];
            } else {
                if (ptr.TryGetValue(t, tn) == false) {
                    tn = _first[t];
                } else {
                    tn=ptr.GetValue(t);//TODO: 不太清楚 java out参数 怎么做？ 
                }
            }
            if (tn != null) {
                if (tn.End) {
                    return true;
                }
            }
            ptr = tn;
        }
        return false;
    }

    public String Replace(String text){
        return Replace(text,'*');
    }

    public String Replace(String text,Character replaceChar)
    {
        StringBuilder result = new StringBuilder(text);

        TrieNode ptr = null;
        for (int i = 0; i < text.length(); i++) {
            Character t=text.charAt(i);
            TrieNode tn=null;
            if (ptr == null) {
                tn = _first[t];
            } else {
                if (ptr.TryGetValue(t,   tn) == false) {
                    tn = _first[t];
                }else {
                    tn=ptr.GetValue(t);//TODO: 不太清楚 java out参数 怎么做？ 
                }
            }
            if (tn != null) {
                if (tn.End) {
                    int maxLength = tn.Results.get(0).length();
                    int start = i + 1 - maxLength;
                    for (int j = start; j <= i; j++) {
                        result.setCharAt(j, replaceChar);
                    }
                }
            }
            ptr = tn;
        }
        return result.toString();
    }


}

File: golang/ToolGood/Words/StringSearch.go
package Words

import (
	. "./internals"
)

type StringSearch struct {
	_first    []*TrieNode2
	_keywords []string
}

func NewStringSearch() *StringSearch {
	return &StringSearch{
		_first: make([]*TrieNode2, 0),
	}
}

func (this *StringSearch) SetKeywords(keywords []string) {
	this._keywords = keywords

	root := NewTrieNode()
	allNodeLayers := make(map[int]*TrieNodes, 0)

	for r, p := range keywords {
		length := len(p)
		if length > 0 {
			var nd *TrieNode
			nd = root
			for i, ch := range p {
				nd = nd.Add(int(ch))
				if nd.Layer == 0 {
					nd.Layer = i + 1
					if trieNodes, ok := allNodeLayers[nd.Layer]; ok {
						trieNodes.Items = append(trieNodes.Items, nd)
					} else {
						trieNodes = NewTrieNodes()
						allNodeLayers[nd.Layer] = trieNodes
						trieNodes.Items = append(trieNodes.Items, nd)
					}
				}
			}
			nd.SetResults(r)
		}
	}

	allNode := make([]*TrieNode, 1)
	allNode[0] = root
	for i := 0; i < len(allNodeLayers); i++ {
		nodes := allNodeLayers[i+1].Items
		for j := 0; j < len(nodes); j++ {
			allNode = append(allNode, nodes[j])
		}
	}

	for i := 1; i < len(allNode); i++ {
		var nd *TrieNode
		var r *TrieNode

		nd = allNode[i]
		nd.Index = i
		r = nd.Parent.Failure
		c := nd.Char
		for {
			if r != nil {
				if _, ok := r.M_values[c]; ok {
					break
				} else {
					r = r.Failure
				}
			} else {
				break
			}
		}

		if r == nil {
			nd.Failure = root
		} else {
			nd.Failure = r.M_values[c]
			for j := 0; j < len(nd.Failure.Results); j++ {
				nd.SetResults(nd.Failure.Results[j])
			}
		}
	}
	root.Failure = root

	allNode2 := make([]*TrieNode2, len(allNode))
	for i := 0; i < len(allNode); i++ {
		allNode2[i] = NewTrieNode2()
	}
	for i := 0; i < len(allNode); i++ {
		oldNode := allNode[i]
		newNode := allNode2[i]

		for key, val := range oldNode.M_values {
			var index = val.Index
			newNode.Add(int32(key), allNode2[index])
		}
		for j := 0; j < len(oldNode.Results); j++ {
			newNode.SetResults(oldNode.Results[j])
		}
		oldNode = oldNode.Failure
		for oldNode != root {
			for key, val := range oldNode.M_values {
				if newNode.HasKey(int32(key)) == false {
					var index = val.Index
					newNode.Add(int32(key), allNode2[index])
				}
			}
			for j := 0; j < len(oldNode.Results); j++ {
				newNode.SetResults(oldNode.Failure.Results[j])
			}
			oldNode = oldNode.Failure
		}
	}

	first := make([]*TrieNode2, 0x10000)
	for key, val := range allNode2[0].M_values {
		first[key] = val
	}
	this._first = first
}

func (this *StringSearch) FindFirst(text string) string {
	var ptr *TrieNode2
	for _, t := range text {
		var tn *TrieNode2
		if ptr == nil {
			tn = this._first[t]
		} else {
			var b bool
			b, tn = ptr.TryGetValue(t)
			if b == false {
				tn = this._first[t]
			}
		}
		if tn != nil {
			if tn.End == true {
				return this._keywords[tn.Results[0]]
			}
		}
		ptr = tn
	}
	return ""
}

func (this *StringSearch) FindAll(text string) []string {
	list := make([]string, 0)
	var ptr *TrieNode2
	for _, t := range text {
		var tn *TrieNode2
		if ptr == nil {
			tn = this._first[t]
		} else {
			var b bool
			b, tn = ptr.TryGetValue(t)
			if b == false {
				tn = this._first[t]
			}
		}
		if tn != nil {
			if tn.End == true {
				for _, item := range tn.Results {
					list = append(list, this._keywords[item])
				}
			}
		}
		ptr = tn
	}
	return list
}

func (this *StringSearch) ContainsAny(text string) bool {
	var ptr *TrieNode2
	for _, t := range text {
		var tn *TrieNode2
		if ptr == nil {
			tn = this._first[t]
		} else {
			var b bool
			b, tn = ptr.TryGetValue(t)
			if b == false {
				tn = this._first[t]
			}
		}
		if tn != nil {
			if tn.End == true {
				return true
			}
		}
		ptr = tn
	}
	return false
}

func (this *StringSearch) Replace(text string, replaceChar int32) string {
	result := []rune(text)
	var ptr *TrieNode2
	var i int
	for _, t := range text {
		var tn *TrieNode2
		if ptr == nil {
			tn = this._first[t]
		} else {
			var b bool
			b, tn = ptr.TryGetValue(t)
			if b == false {
				tn = this._first[t]
			}
		}
		if tn != nil {
			if tn.End == true {
				maxLength := len([]rune(this._keywords[tn.Results[0]]))
				start := i + 1 - maxLength
				for j := start; j <= i; j++ {
					result[j] = replaceChar
				}
			}
		}
		ptr = tn
		i++
	}
	return string(result)
}
